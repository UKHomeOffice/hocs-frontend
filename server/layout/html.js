const serialize = require('serialize-javascript');

let assets = {
    js: [],
    css: []
};

const render = ({
                    assetPath = '/public',
                    css = [],
                    js = [],
                    react = 'main',
                    title = 'My React Application',
                    propositionHeader = '',
                    markup = '',
                    props = {},
                    clientSide = true
                }) => {
    return (`<!DOCTYPE html>
                <html>
                <head>
                    <meta charSet="utf-8"/>
                    <title>${title || propositionHeader}</title>                            
                
                    <link rel="shortcut icon" href="${assetPath}/images/favicon.ico" type="image/x-icon"/>
                    <link rel="stylesheet" media="screen" href="${assetPath}/stylesheets/govuk-template.css?0.23.0"/>
                    <link rel="stylesheet" media="print" href="${assetPath}/stylesheets/govuk-template-print.css?0.23.0"/>
                    <link rel="stylesheet" media="all" href="${assetPath}/stylesheets/fonts.css?0.23.0"/>
                    <link rel="mask-icon" href="${assetPath}/images/gov.uk_logotype_crown.svg" color="#0b0c0c">
                    <link rel="apple-touch-icon" sizes="180x180" href="${assetPath}/images/apple-touch-icon-180x180.png?0.23.0">
                    <link rel="apple-touch-icon" sizes="167x167" href="${assetPath}/images/apple-touch-icon-167x167.png?0.23.0">
                    <link rel="apple-touch-icon" sizes="152x152" href="${assetPath}/images/apple-touch-icon-152x152.png?0.23.0">
                    <link rel="apple-touch-icon" href="${assetPath}/images/apple-touch-icon.png?0.23.0">            
                    ${css.map(style => `<link rel="stylesheet" media="screen" href="${assetPath}/styles/${assets.css[style]}" />`).join('\n')}
                    ${clientSide ? js.map(script => `<script src="${assetPath}/js/${assets.js[script]}"></script>`).join('\n') : ''}
                    ${clientSide ? `<script>window.__INITIAL_DATA__ = ${serialize(props)}</script>` : ''}
                    
                    <meta name="viewport" content="width=device-width, initial-scale=1"/>
                    <meta property="og:image" content=${assetPath}/images/opengraph-image.png?0.23.0/>                    
                </head>
                
                <body>
                    <script>
                        document.body.className = ((document.body.className) ? 
                        document.body.className + ' js-enabled' :
                        'js-enabled');
                    </script>
                    
                    <div id="skiplink-container">
                        <div>
                            <a href="#content" class="skiplink">Skip to main content</a>
                        </div>
                    </div>
                    
                    <div id="global-cookie-message">
                        <p>
                            GOV.UK uses cookies to make the site simpler.
                            <a href="https://www.gov.uk/help/cookies">Find out more about cookies</a>
                        </p>
                    </div>
                        
                    <div id="app">${markup}</div>
                    ${clientSide ? `<script src="${assetPath}/js/${assets.js[react]}" defer></script> `: ''}
                    
                    <div id="global-app-error" className="app-error hidden"></div>
                
                    <script src=${assetPath}/javascripts/govuk-template.js></script>
                    <script src="${assetPath}/javascripts/ie.js"></script>
                    <script>
                        if (typeof window.GOVUK === 'undefined') 
                        document.body.className = document.body.className.replace('js-enabled', '');
                    </script>
                </body>
                </html>`);
};

module.exports = {
    use: (a) => assets = a,
    render
};