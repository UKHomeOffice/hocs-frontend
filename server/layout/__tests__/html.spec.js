import html from "../html";

html.use({
    "js": {
        "main": "js/main.js",
        "first": "js/first.js",
        "second": "js/second.js",
        "myBundle": "js/myBundle.js"
    },
    "css": {
        "first": "styles/first.css"
    }
});

describe('HTML module', () => {
    it('should return the default html when no additional configuration is passed', () => {
        const configuration = {};
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should include the server rendered markup when passed', () => {
        const configuration = {
            markup: '<--- REACT APPLICATION --->'
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should not contain the react bundle js when clientside is set to false', () => {
        const configuration = {
            clientside: false
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should include additional CSS and JS when passed', () => {
        const configuration = {
            js: ['first', 'second'],
            css: ['first']
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should include the appropriate react application bundle when passed', () => {
        const configuration = {
            react: 'myBundle'
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should default to proposition header when no title is passed', () => {
        const configuration = {
            propositionHeader: 'Proposition Header',
            title: null
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should include the correct asset path when passed', () => {
        const configuration = {
            assetPath: 'some/public/folder/that/contains/my/assets'
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
    it('should include application props in the head when passed', () => {
        const configuration = {
            props: {
                first: 'value',
                second: 'value'
            }
        };
        const markup = html.render(configuration);
        expect(markup).toBeDefined();
        expect(markup).toMatchSnapshot();
    });
});