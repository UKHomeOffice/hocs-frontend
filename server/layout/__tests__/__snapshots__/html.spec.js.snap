// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTML module should default to proposition header when no title is passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>Proposition Header</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/main.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should include additional CSS and JS when passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    <link rel=\\"stylesheet\\" media=\\"screen\\" href=\\"/public/styles/styles/first.css\\" />
    <script defer src=\\"/public/js/js/first.js\\"></script>
<script defer src=\\"/public/js/js/second.js\\"></script>
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/main.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should include application props in the head when passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {\\"first\\":\\"value\\",\\"second\\":\\"value\\"}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/main.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should include the appropriate react application bundle when passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/myBundle.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should include the correct asset path when passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"some/public/folder/that/contains/my/assets/js/js/main.js\\"></script>
    <script defer src=\\"some/public/folder/that/contains/my/assets/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"some/public/folder/that/contains/my/assets/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should include the server rendered markup when passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/main.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"><--- REACT APPLICATION ---></div>

</body>

</html>"
`;

exports[`HTML module should not contain the react bundle js when clientside is set to false 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;

exports[`HTML module should return the default html when no additional configuration is passed 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\" class=\\"govuk-template \\">

<head>
    <meta charset=\\"utf-8\\" />
    <title>My React Application</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
    <meta name=\\"theme-color\\" content=\\"#0b0c0c\\" />

    <link rel=\\"shortcut icon\\" href=\\"/public/assets/images/favicon.ico\\" type=\\"image/x-icon\\" />
    <link rel=\\"mask-icon\\" href=\\"/public/assets/images/govuk-mask-icon.svg\\" color=\\"#0b0c0c\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"180x180\\" href=\\"/public/assets/images/govuk-apple-touch-icon-180x180.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"167x167\\" href=\\"/public/assets/images/govuk-apple-touch-icon-167x167.png\\">
    <link rel=\\"apple-touch-icon\\" sizes=\\"152x152\\" href=\\"/public/assets/images/govuk-apple-touch-icon-152x152.png\\">
    <link rel=\\"apple-touch-icon\\" href=\\"/public/assets/images/govuk-apple-touch-icon.png\\">
    
    
    <script defer >window.__INITIAL_DATA__ = {}</script>
     <script defer>
        document.body.className = ((document.body.className) ? document.body.className + ' js-enabled' : 'js-enabled');
    </script>
    <script defer src=\\"/public/js/js/main.js\\"></script>
    <script defer src=\\"/public/all.js\\"></script>
    <script defer>
        window.GOVUKFrontend.initAll()
    </script>
    <meta property=\\"og:image\\" content=\\"/public/assets/images/govuk-opengraph-image.png\\">
</head>

<body class=\\"govuk-template__body \\">

    <div id=\\"app\\"></div>

</body>

</html>"
`;
